### ISSUE TITLE
checkpoint callback not monitor correct value

### ISSUE THREAD

**Issue by xiadingZ** [PyTorchLightning/pytorch-lightning#4067]
pytorch-lightning version 1.0.0rc2

This is my checkpoint callback:
```
    checkpoint_callback = ModelCheckpoint(
        monitor='val_acc',
        mode='max',
    )
```
this is my `validation_step`:
```
        ...
        self.log('val_loss', loss, on_epoch=True, sync_dist=True)
        self.log('val_pacc', part_acc, on_epoch=True, sync_dist=True)
        self.log('val_acc', vid_acc, on_epoch=True, prog_bar=True, sync_dist=True)



        return {'val_loss': loss, "val_pacc": part_acc, "val_acc": vid_acc}
```
my tensorboard shows my max `val_acc` is at epoch 39, but checkpoint saved model is `epoch=67`.

I run two experiments, but all saved models are at wrong epoch. 
Maybe monitor value is not `sync_dist`ed, but tensorboard's value is already `sync_dist`?

**Comment by awaelchli** [comment]
Could you verify this with the template model for bug reports here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py
and paste the code here?

You can also set verbose=True in ModelCheckpoint to see the values that the callback receives when saving the checkpoint

**Comment by awaelchli** [comment]
Note, when running multi-GPU, CheckpointCallback only saves on rank 0.

**Comment by xcliang2** [comment]
@awaelchli ，i verify this with the template [model(),](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py) and my run_test code is

        def validation_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('val_loss', loss)
            self.log('test_batch_idx', batch_idx)
            return {"x": loss}

    # fake data
    train_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    val_data = torch.utils.data.DataLoader(RandomDataset(32, 64))
    test_data = torch.utils.data.DataLoader(RandomDataset(32, 64))

    save_model_callback = callbacks.ModelCheckpoint(
        filepath=os.getcwd() + '/models/{epoch:002d}-{val_loss:.4f}--{test_batch_idx:0.0f}',
        save_last=True,
        save_top_k=1,
        monitor='val_loss',
        mode='min',
        verbose=True,
    )

    # model
    model = TestModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1.0,
        limit_val_batches=1.0,
        max_epochs=10,
        weights_summary=None,
        checkpoint_callback=save_model_callback
    )
    trainer.fit(model, train_data, val_data)
    trainer.test(test_dataloaders=test_data)

but when i check my log, i found out the value test_batch_idx is always 63, like this

![QQ图片20201015114623](https://user-images.githubusercontent.com/31505323/96074857-464d5800-0edc-11eb-8187-837ed49bc300.png)

but model's name is correct like this
![微信图片_20201015115205](https://user-images.githubusercontent.com/31505323/96075106-e6a37c80-0edc-11eb-83a5-130c695c7da6.png)
 I think this the same bug as discussed in #4141

**Comment by awaelchli** [comment]
![image](https://user-images.githubusercontent.com/5495193/96341245-55c5d000-1099-11eb-8f38-9c696fa2f0e6.png)

`self.log` computes the average of the metric at the end of the epoch. So when you log the test_batch_idx, it will always compute the average of all values 0,1,2,3,4....,n where n is the number of test_steps.
In the attached image you see that when I change to `self.log('test_batch_idx', batch_idx, on_step=True)`, I get individual plots for each epoch, and the batch_idx counter increases as it should. 

if you want the step in your filename, I recommend to use `self.global_step` instead. 

Let me know if you have any further questions.
